Issue #5850 support Accelerate your network share hosted Package Source

nschwerzler opened the pull request:
Link: https://github.com/NuGet/Home/issues/5850
 Regression: No

Fix: Added functionality to nuget.exe pack command to support network share hosted package source requirements

Details:

Added two arguments to the pack command to support Accelerate your network share hosted Package Source
- OutputSHA512PackageHash: Creates the [packageName].nupkg.sha512 file that contains the package hash
- OutputResolvedNuSpec: Writes the resolved nuspec file to the package output directory which contains the same contents as the nuspec in the nupkg. If you attempt to use this option where the input nuspec is the same as the output resolved nuspec, will throw so user does not overwrite the original accidently.

Testing/Validation

Tests Added: Yes

Added 3 NuGetPackCommandTests
- PackCommand_OutputResolvedNuSpecFileAttemptToOverwriteOriginal: Verifies nuget.exe pack will not overwrite the users original nuspec when writing a resolved nuspec file to the package output directory
- PackCommand_OutputResolvedNuSpecFile: Verifies nuget.exe pack will output the resolved nuspec file with token replacements and is identical as the one in the nupkg
- PackCommand_OutputSHA512PackageHash: Verifies nuget.exe pack will output a sha512 package hash file

 Validation done:
Verified new added tests pass inside VS test explorer UI
Built product in visual studio and build.ps1 and verified passed
Looked at test results output for build.ps1 and verified tests passed

## Bug
Link: Link_to_issue  
Regression: Yes/No  
If Regression then when did it last work:   
If Regression then how are we preventing it in future:   

## Fix
Details: Details_about the_fix  

## Testing/Validation
Tests Added: Yes/No  
Reason for not adding tests:  
Validation done:

nschwerzler committed:
Bug 5850

Link: https://github.com/NuGet/Home/issues/5850
 Regression: No

Fix: Added functionality to nuget.exe pack command to support network share hosted package source requirements

Details:

Added two arguments to the pack command to support Accelerate your network share hosted Package Source
- OutputSHA512PackageHash: Creates the [packageName].nupkg.sha512 file that contains the package hash
- OutputResolvedNuSpec: Writes the resolved nuspec file to the package output directory which contains the same contents as the nuspec in the nupkg. If you attempt to use this option where the input nuspec is the same as the output resolved nuspec, will throw so user does not overwrite the original accidently.

Testing/Validation

Tests Added: Yes

Added 3 NuGetPackCommandTests
- PackCommand_OutputResolvedNuSpecFileAttemptToOverwriteOriginal: Verifies nuget.exe pack will not overwrite the users original nuspec when writing a resolved nuspec file to the package output directory
- PackCommand_OutputResolvedNuSpecFile: Verifies nuget.exe pack will output the resolved nuspec file with token replacements and is identical as the one in the nupkg
- PackCommand_OutputSHA512PackageHash: Verifies nuget.exe pack will output a sha512 package hash file

 Validation done:
Verified new added tests pass inside VS test explorer UI
Built product in visual studio and build.ps1 and verified passed
Looked at test results output for build.ps1 and verified tests passed

dnfclas commented:
@nschwerzler,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by .NET Foundation. We will now review your pull request.
_Thanks,
.NET Foundation Pull Request Bot_

emgarten commented on src/NuGet.Clients/NuGet.CommandLine/Commands/PackCommand.cs:
```diff
@@ -79,6 +79,12 @@ public ICollection<string> Exclude
+        public bool OutputResolvedNuSpec { get; set; }
+
+        [Option(typeof(NuGetCommand), "PackageCommandOutputSHA512PackageHash")]
+        public bool OutputSHA512PackageHash { get; set; }
```
Should these flags be combined? Or should this be a config option instead? 

I'm worried about having too many knobs on pack, there are already a lot of options.

nschwerzler replied:
We could combine them. As far as I know NuGet pack does not support the config file option.

emgarten replied:
Pack could take the config file option, I'm not sure that is the best place for it, just an option. 
@rohit21agrawal

rohit21agrawal replied:
i'd rather have an option like -InstallPackageToV3Feed which takes the path to the feed as a value.

nschwerzler replied:
Will consolidate to one option

rohit21agrawal commented on src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PackageBuilder.cs:
```diff
@@ -314,6 +314,15 @@ public Version MinClientVersion
+        /// <summary>
+        /// Gets the package manifest which is available after WriteManifest
+        /// </summary>
+        public Manifest PackageManifest
```
why do you need this as a property?

nschwerzler replied:
Removing, not needed anymore

rohit21agrawal commented on src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/Manifest.cs:
```diff
@@ -51,6 +51,11 @@ public Manifest(ManifestMetadata metadata, ICollection<ManifestFile> files)
         /// <summary>
+        /// Gets the resolved manifest XDocument populated after save
+        /// </summary>
+        public XDocument ResolvedManifest { get; private set; }
```
same here, why is this needed as a property?

nschwerzler replied:
Removing, not needed anymore

rohit21agrawal commented:
if we are doing this, we also need to do feature parity for dotnet.exe pack as well. @rrelyea  thoughts?

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:
```diff
@@ -131,10 +131,69 @@ public PackageArchiveReader BuildPackage(PackageBuilder builder, string outputPa
+        /// <param name="packageBuilder">The package builder</param>
+        private void WriteResolvedNuSpecToPackageOutputDirectory(string packageOutputPath, PackageBuilder packageBuilder)
+        {
+            string packageOutputDirectory = Path.GetDirectoryName(packageOutputPath);
```
we are doing something similar here:
https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Commands/MSBuildProjectFactory.cs#L90-L108

i'd rather have a similar implementation at both places. you just need to pass null for the FileCollection to Manifest constructor

nschwerzler replied:
Nice, thank you for link. Updating to use the same implementation in the linked code.

nschwerzler replied:
Resolved

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:
```diff
@@ -131,10 +131,69 @@ public PackageArchiveReader BuildPackage(PackageBuilder builder, string outputPa
+        /// <param name="packageBuilder">The package builder</param>
+        private void WriteSHA512PackageHash(string packageOutputPath, PackageBuilder packageBuilder)
+        {
+            string packageOutputDirectory = Path.GetDirectoryName(packageOutputPath);
```
@emgarten there must already be a helper function that does this , right? in install/restore? can you point nick to it?

rohit21agrawal commented:
i would also like to discuss how to handle symbols nupkgs here.

rohit21agrawal reviewed (changes requested):
left comments for feedback

nschwerzler committed:
Iteration 2 based on code review feedback
- Combined two args for creating feed into one
- Removed two properties for generating manifest and use existing code pattern.

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:
```diff
@@ -131,10 +131,63 @@ public PackageArchiveReader BuildPackage(PackageBuilder builder, string outputPa
+        private void WriteResolvedNuSpecToPackageOutputDirectory(string packageOutputPath, PackageBuilder packageBuilder)
+        {
+            string packageOutputDirectory = Path.GetDirectoryName(packageOutputPath);
+            string resolvedNuSpecOutputPath = Path.Combine(packageOutputDirectory, packageBuilder.Id + PackagingConstants.ManifestExtension);
```
could you use this to get outputPath: https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs#L978 
modify the function to accept a parameter which decides whether or not to append version. you can use it for the sha512 file too

nschwerzler replied:
Resolved.

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:
```diff
@@ -131,10 +131,63 @@ public PackageArchiveReader BuildPackage(PackageBuilder builder, string outputPa
+            }
+
+            Manifest manifest = new Manifest(new ManifestMetadata(packageBuilder), null);
+            using (Stream stream = new FileStream(resolvedNuSpecOutputPath, FileMode.Create))
```
have you checked behavior about what happens if file already exists?

nschwerzler replied:
No errors when file already exists and refreshes with new file.

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:
```diff
@@ -131,10 +131,63 @@ public PackageArchiveReader BuildPackage(PackageBuilder builder, string outputPa
                 PrintVerbose(outputPath, builder);
             }
 
+            if (_packArgs.InstallPackageToV3Feed)
```
the normal nupkg should not be generated.

nschwerzler replied:
Resolved

rohit21agrawal commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPackCommandTest.cs:
```diff
@@ -2085,6 +2085,181 @@ public void PackCommand_NuspecFileWithTokens()
+                // Verify the zip file has the resolved nuspec
+                XDocument nuspecZipXml = null;
+                var nupkgPath = Path.Combine(workingDirectory, "packageA.1.2.3.4.nupkg");
+                using (var zip = new ZipArchive(File.OpenRead(nupkgPath)))
```
we have PackageArchiveReader and NuspecReader to do these tasks.
https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs#L877-L880

nschwerzler replied:
Resolved

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/CommandArgs/PackArgs.cs:
```diff
@@ -22,6 +22,7 @@ public class PackArgs
         public ILogger Logger { get; set; }
         public LogLevel LogLevel { get; set; }
         public bool IncludeReferencedProjects { get; set; }
+        public bool InstallPackageToV3Feed { get; set; }
```
@rrelyea how about InstallPackageToOutputPath ? I propose that the flag (which is a boolean, doesn't accept any value)does the following:
1) Installs the package to the output path as specified by user
2) Prevents the normal nupkg (and symbols nupkg) from being generated.
3) Logs where the package was installed

nschwerzler replied:
Resolved. Updated name. Normal nupkg and symbols are no longer generated.

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:
```diff
@@ -131,10 +131,63 @@ public PackageArchiveReader BuildPackage(PackageBuilder builder, string outputPa
+                WriteSHA512PackageHash(outputPath, builder);
+            }
+
             _packArgs.Logger.LogMinimal(String.Format(CultureInfo.CurrentCulture, Strings.Log_PackageCommandSuccess, outputPath));
```
the logging needs to be updated too if you are installing the package

nschwerzler replied:
Not installing the package, so this is not needed unless I am misunderstanding the request.

rohit21agrawal replied:
by installing the package I mean when InstallPackageToV3Feed is set

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:
```diff
@@ -131,10 +131,63 @@ public PackageArchiveReader BuildPackage(PackageBuilder builder, string outputPa
+                throw new PackagingException(
+                    string.Format(
+                        CultureInfo.CurrentCulture,
+                        "Unable to output resolved nuspec file because it would overwrite the original at '{0}'",
```
add this entry to the corresponding Strings.resx file so it can be localised.

nschwerzler replied:
Resolved

rohit21agrawal commented:
>             if (_packArgs.Symbols)

you also need to check if InstallPackageToOutputPath is specified and prevent this from happening

---
Refers to: src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs:739 in b2dde82. [](commit_id = b2dde829e3dd9a1ba33efadb4b6e141801638b35, deletion_comment = False)

rohit21agrawal commented:
also change the implementation of PackCommandRunner.BuildSymbolsPackage so it no-ops if InstallPackageToOutputPath is set

nschwerzler committed:
Issue #5850 support Accelerate your network share hosted Package Source. Iteration 3. https://github.com/NuGet/Home/issues/5850
